db/snapshot-ctl.cc

```diff
@@ -174,51 +181,41 @@ snapshot_ctl::get_snapshot_details() {
-            auto local_snapshots = make_lw_shared<snapshot_map>();
-            return parallel_for_each(db.get_column_families(), [local_snapshots] (auto& cf_pair) {
-                return cf_pair.second->get_snapshot_details().then([uuid = cf_pair.first, local_snapshots] (auto map) {
-                    for (auto&& snap_map: map) {
-                        auto [it, ignored] = local_snapshots->try_emplace(snap_map.first);
-                        it->second.emplace(uuid, snap_map.second);
-                    }
-                    return make_ready_future<>();
-                });
-            }).then([local_snapshots] {
-                return make_ready_future<snapshot_map>(std::move(*local_snapshots));
-            });
-        }).then([this] (snapshot_map&& map) {
-            std::unordered_map<sstring, std::vector<snapshot_ctl::snapshot_details>> result;
-            for (auto&& pair: map) {
-                std::vector<snapshot_ctl::snapshot_details> details;
-
-                for (auto&& snap_map: pair.second) {
-                    auto& cf = _db.local().find_column_family(snap_map.first);
-                    details.push_back({ snap_map.second.live, snap_map.second.total, cf.schema()->cf_name(), cf.schema()->ks_name() });
-                }
-                result.emplace(pair.first, std::move(details));
+    co_return co_await run_snapshot_list_operation([this] () -> future<details_result> {
+        snapshot_map map = co_await _db.map_reduce(snapshot_reducer(), [] (database& db) -> future<snapshot_map> {
+            co_return co_await db.get_snapshot_details();
+        });
+
+        details_result result;
+        for (auto&& pair: map) {
+            std::vector<snapshot_ctl::snapshot_details> details;
+
+            for (auto&& snap_map: pair.second) {
+                details.push_back(snap_map.second);
             }
+            result.emplace(pair.first, std::move(details));
+        }
 
-            return make_ready_future<std::unordered_map<sstring, std::vector<snapshot_ctl::snapshot_details>>>(std::move(result));
-        });
+        co_return result;
     });
 }
 
 future<int64_t> snapshot_ctl::true_snapshots_size() {
-    return run_snapshot_list_operation([this] {
-        return _db.map_reduce(adder<int64_t>(), [] (database& db) {
-            return do_with(int64_t(0), [&db] (auto& local_total) {
-                return parallel_for_each(db.get_column_families(), [&local_total] (auto& cf_pair) {
-                    return cf_pair.second->get_snapshot_details().then([&local_total] (auto map) {
-                        for (auto&& snap_map: map) {
-                            local_total += snap_map.second.live;
+    co_return co_await run_snapshot_list_operation([this] () -> future<int64_t> {
+        co_return co_await _db.map_reduce(adder<int64_t>(), [] (database& db) -> future<int64_t> {
+            co_return co_await do_with(int64_t(0), [&db] (auto& local_total) -> future<int64_t> {
```

Comment:
Since you turned this into a coroutine, you don't need the `do_with` anymore.

Reply:
ah, right